client/windows/QuickRecruitmentWindow.cpp

```diff
@@ -0,0 +1,157 @@
+			pos.x += 108;
+		}
+	}
+	totalCost = std::make_shared<CreatureCostBox>(Rect((this->pos.w/2)-45, pos.y + 260, 97, 74), "");
+}
+
+void QuickRecruitmentWindow::initWindow()
+{
+	pos.x = 238;
+	pos.y = 45;
+	pos.w = 332;
+	pos.h = 461;
+	int creaturesAmount = getAvailableCreatures();
+	if(creaturesAmount > 3)
+	{
+		pos.w += 108 * (creaturesAmount - 3);
+		pos.x -= 55 * (creaturesAmount - 3);
+	}
+	backgroundTexture = std::make_shared<CFilledTexture>("DIBOXBCK.pcx", Rect(0, 0, pos.w, pos.h));
+}
+
+void QuickRecruitmentWindow::maxAllCards(std::vector<std::shared_ptr<CreaturePurhaseCard> > cards)
+{
+	auto allAvailableResources = LOCPLINT->cb->getResourceAmount();
+	for(auto i : boost::adaptors::reverse(cards))
+	{
+		si32 maxAmount = i->creature->maxAmount(allAvailableResources);
+		vstd::amin(maxAmount, i->maxamount);
+
+		i->slider->setAmount(maxAmount);
+
+		if(i->slider->getValue() != maxAmount)
+			i->slider->moveTo(maxAmount);
+		else
+			i->sliderMoved(maxAmount);
+
+		i->slider->moveToMax();
+		allAvailableResources -= (i->creature->cost * maxAmount);
+	}
+	maxButton->block(allAvailableResources == LOCPLINT->cb->getResourceAmount());
+}
+
+
+void QuickRecruitmentWindow::purhaseUnits()
+{
+	for(auto selected : cards)
+	{
+		if(selected->slider->getValue())
+		{
+			auto onRecruit = [=](CreatureID id, int count){ LOCPLINT->cb->recruitCreatures(town, town->getUpperArmy(), id, count, selected->creature->level-1); };
+			CreatureID crid =  selected->creature->idNumber;
+			SlotID dstslot = town -> getSlotFor(crid);
+			if(!dstslot.validSlot())
+				continue;
+			onRecruit(crid, selected->slider->getValue());
+		}
+	}
+	close();
+}
+
+int QuickRecruitmentWindow::getAvailableCreatures()
+{
+	int creaturesAmount = 0;
+	for (int i=0; i< GameConstants::CREATURES_PER_TOWN; i++)
+		if(!town->town->creatures.at(i).empty() && !town->creatures.at(i).second.empty() && town->creatures[i].first)
+			creaturesAmount++;
+	return creaturesAmount;
+}
+
+void QuickRecruitmentWindow::updateAllSliders()
+{
+	auto allAvailableResources = LOCPLINT->cb->getResourceAmount();
+	for(auto i : boost::adaptors::reverse(cards))
+		allAvailableResources -= (i->creature->cost * i->slider->getValue());
+	for(auto i : cards)
+	{
+		si32 maxAmount = i->creature->maxAmount(allAvailableResources);
+		vstd::amin(maxAmount, i->maxamount);
+		if(maxAmount < 0)
+			continue;
+		if(i->slider->getValue() + maxAmount < i->maxamount)
+			i->slider->setAmount(i->slider->getValue() + maxAmount);
+		else
+			i->slider->setAmount(i->maxamount);
+		i->slider->moveTo(i->slider->getValue());
+	}
+	totalCost->createItems(LOCPLINT->cb->getResourceAmount() - allAvailableResources);
+	totalCost->set(LOCPLINT->cb->getResourceAmount() - allAvailableResources);
+}
+
+QuickRecruitmentWindow::QuickRecruitmentWindow(const CGTownInstance * townd) : CWindowObject(PLAYER_COLORED | BORDERED), town(townd)
+{
+	OBJ_CONSTRUCTION_CAPTURING_ALL;
```

Comment:
Same issue with smart pointers mixed with DISPOSE flag. If you want to use smart pointers you need modified `OBJ_CONSTRUCTION_CAPTURING_ALL` macro: `CIntObject::DISPOSE` bit must be cleared in defActions.